flowz/test/artifacts/artifacts_test.py

```diff
@@ -0,0 +1,295 @@
+
+    @tt.gen_test
+    def test_wrapped_artifact(self):
+        def deriver(num_arr, num_dict):
+            return OrderedDict([(i, num_dict[i]) for i in num_arr])
+
+        artifact = WrappedArtifact(DerivedArtifact(deriver, self.NUM_ARR, self.NUM_DICT))
+
+        tools.assert_false(artifact.exists())
+        value = yield artifact.get()
+        tools.assert_equal(value, self.NUM_ORDERED_DICT)
+        tools.assert_true(artifact.exists())
+        tools.assert_true(artifact.ensure())
+
+    @tt.gen_test
+    def test_transformed_artifact(self):
+        @gen.coroutine
+        def getter():
+            raise gen.Return(self.NUM_ORDERED_DICT)
+
+        def deriver(num_arr, num_dict):
+            return OrderedDict([(i, num_dict[i]) for i in num_arr])
+
+        def transformer(orig_dict):
+            return OrderedDict([(i, orig_dict[i]) for i in reversed(orig_dict.keys())])
+
+        # Try with an ExtantArtifact
+        artifact = TransformedArtifact(ExtantArtifact(getter), transformer)
+
+        tools.assert_true(artifact.exists())
+        tools.assert_true(artifact.ensure())
+
+        value = yield artifact.get()
+
+        tools.assert_equal(value, self.NUM_REVERSED_DICT)
+
+        # Try with a DerivedArtifact
+        artifact = TransformedArtifact(DerivedArtifact(deriver, self.NUM_ARR, self.NUM_DICT),
+                                       transformer)
+
+        tools.assert_false(artifact.exists())
+
+        value = yield artifact.get()
+
+        tools.assert_equal(value, self.NUM_REVERSED_DICT)
+        tools.assert_true(artifact.exists())
+        tools.assert_true(artifact.ensure())
+
+    @tt.gen_test
+    def test_keyed_artifact(self):
+        def deriver(key, num_dict):
+            return num_dict[key]
+
+        key = 1
+        artifact = KeyedArtifact(key, DerivedArtifact(deriver, key, self.NUM_DICT))
+
+        tools.assert_equal(artifact[0], key)
+        tools.assert_equal(artifact[1], artifact)
+        tools.assert_equal(artifact['key'], key)
+        tools.assert_raises(KeyError, artifact.__getitem__, 'spaz')
+
+        for (a,b) in zip((key, artifact), iter(artifact)):
+            tools.assert_equal(a, b)
+
+        tools.assert_false(artifact.exists())
+
+        value = yield artifact.get()
+
+        tools.assert_equal(value, 'one')
+        tools.assert_true(artifact.exists())
+        tools.assert_true(artifact.ensure())
+
+    @tt.gen_test
+    def test_keyed_artifact_transform(self):
+        def deriver(key, num_dict):
+            return num_dict[key]
+
+        def rderiver(num_dict, val):
+            for (k, v) in num_dict.iteritems():
+                if v == val:
+                    return k
+            return None
+
+        key = 1
+        artifact = KeyedArtifact(key, DerivedArtifact(deriver, key, self.NUM_DICT))
+        artifact2 = artifact.transform(rderiver, self.NUM_DICT)
+
+        key2 = yield artifact2.get()
+        tools.assert_equal(key, key2)
```

Comment:
What is the expectation of `transform()` on a `KeyedArtifact` as opposed to other more general artifacts?

Do you expect that the result of `transform()` will _also_ be a `KeyedArtifact`, with the same key as the original?

I wonder if we should have that expectation.